darkstat.8
.\"
.\" darkstat 3
.\" Copyright 2001-2010, Emil Mikulic.
.\"
.\" You may use, modify and redistribute this file under the terms of the
.\" GNU General Public License version 2. (see COPYING.GPL)
.\"
.TH darkstat 8 "September 2008" "darkstat 3"
.SH NAME
darkstat \- network statistics gatherer
.\"
.SH SYNOPSIS
.B darkstat
[
.BI \-i " interface"
] [
.BI \-r " file"
] [
.BI \-\-snaplen " bytes"
] [
.BI \-\-pppoe
] [
.BI \-\-syslog
] [
.BI \-\-verbose
] [
.BI \-\-no\-daemon
] [
.BI \-\-no\-promisc
] [
.BI \-\-no\-dns
] [
.BI \-\-no\-macs
] [
.BI \-\-no\-lastseen
] [
.BI \-p " port"
] [
.BI \-b " bindaddr"
] [
.BI \-f " filter"
] [
.BI \-l " network/netmask"
] [
.BI \-\-chroot " dir"
] [
.BI \-\-user " username"
] [
.BI \-\-daylog " filename"
] [
.BI \-\-import " filename"
] [
.BI \-\-export " filename"
] [
.BI \-\-pidfile " filename"
] [
.BI \-\-hosts\-max " count"
] [
.BI \-\-hosts\-keep " count"
] [
.BI \-\-ports\-max " count"
] [
.BI \-\-ports\-keep " count"
] [
.BI \-\-highest\-port " port"
] [
.BI \-\-hexdump
]
.\"
.SH DESCRIPTION
.I darkstat
is a packet sniffer that runs as a background process,
gathers all sorts of statistics about network usage,
and serves them over HTTP.

All settings are passed on the commandline.
.\"
.SH OPTIONS
.\"
.TP
.BI \-i " interface"
Capture traffic on the specified network interface.
This is the only mandatory commandline argument.
.\"
.TP
.BI \-r " file"
Instead of capturing live traffic, read it from a
.BR pcap (3)
capture file.
This is only useful for development and benchmarking.
The
.BI \-r
and
.BI \-i
arguments are mutually exclusive.
.\"
.TP
.BI \-\-snaplen " bytes"
How many bytes to capture from the start of each packet.
You should not need to specify this;
\fIdarkstat\fR will calculate it automatically.
.\"
.TP
.BI \-\-pppoe
Don't use this.

Instead, capture on the tunnel interface that your PPPoE software
provides, for example \fBtun0\fR on \fIFreeBSD\fR, \fBpppoe0\fR on
\fIOpenBSD\fR or \fINetBSD\fR.

If you really must, you can capture on an Ethernet interface and pass
this argument to have \fIdarkstat\fR decode PPPoE frames and ignore
everything else.
Make sure you also specify your local address with the \fB\-l\fR
argument!
.\"
.TP
.BI \-\-syslog
Errors, warnings, and verbose messages will go to \fBsyslog\fR (facility
daemon, priority debug) instead of \fBstderr\fR.

On some systems, these messages end up in \fB/var/log/debug\fR
by default.
.\"
.TP
.BI \-\-verbose
Produce more verbose debugging messages.
.\"
.TP
.BI \-\-no\-daemon
Do not detach from the controlling terminal;
stay in the foreground.
.\"
.TP
.BI \-\-no\-promisc
Do not use promiscuous mode to capture.
Note that an interface may already be in promiscuous mode, or may later
enter promiscuous mode, due to circumstances beyond \fIdarkstat\fR's control.
If this is a problem, use \fB\-f\fR to specify an appropriate
.BR bpf (4)
filter.
.\"
.TP
.BI \-\-no\-dns
Do not resolve IPs to host names.
This can significantly reduce memory footprint on small systems
as an extra process is created for DNS resolution.
.\"
.TP
.BI \-\-no\-macs
Do not display MAC addresses in the hosts table.
.\"
.TP
.BI \-\-no\-lastseen
Do not display the last seen time in the hosts table.
.\"
.TP
.BI \-p " port"
Bind the web interface to the specified port.
The default is 667.
.\"
.TP
.BI \-b " bindaddr"
Bind the web interface to the specified address.
The default is to listen on all interfaces.
.\"
.TP
.BI \-f " filter"
Use the specified filter expression when capturing traffic.
The filter syntax is beyond the scope of this manual page;
please refer to the
.BR tcpdump (1)
documentation.
.\"
.TP
.BI \-l " network/netmask"
Define a "local network" according to the network and netmask addresses.
All traffic entering or leaving this network will be graphed, as opposed
to the default behaviour of only graphing traffic to and from the local
host.
.RS

The rule is that if \fBip_addr & netmask == network\fR,
then that address is considered local.
See the usage example below.
.RE
.\"
.TP
.BI \-\-chroot " dir"
Force \fIdarkstat\fR to \fBchroot()\fR into the specified directory.
Without this argument, a default directory will be used, which is
determined at build time.
Usually \fI/var/empty\fR or \fI/var/lib/empty\fR.
.RS

For security reasons, this directory should be empty, and the user that
\fIdarkstat\fR is running as should not have write access to it.

However, if you wish to use \fB\-\-daylog\fR or \fB\-\-export\fR,
\fIdarkstat\fR will need write access to the chroot.
If you are uncomfortable with the security implications, don't
use any functionality that requires write access.
.RE
.\"
.TP
.BI \-\-user " username"
Force \fIdarkstat\fR to drop privileges to the \fBuid\fR and \fBgid\fR of
the specified user.
Without this argument, a default value will be used, which is set at
build time.
Usually \fBnobody\fR.
.RS

For security reasons, this should not be \fBroot\fR.
.RE
.\"
.TP
.BI \-\-daylog " filename"
.RS
Log daily traffic statistics into the named file, relative to the
chroot directory.
If you wish to use \fB\-\-daylog\fR, you must first specify a
\fB\-\-chroot\fR directory, and it must be writeable by the
\fIdarkstat\fR user.
A writeable chroot has security implications; if you are uncomfortable
with this, do not use the \fB\-\-daylog\fR functionality.

If the daylog argument is not specified, no logging is performed.

The daylog format is:

localtime|time_t|bytes_in|bytes_out|pkts_in|pkts_outs

Lines starting with a # are comments stating when logging started and
stopped.
.RE
.\"
.TP
.BI \-\-import " filename"
Upon starting, import a \fIdarkstat\fR database from the named file,
relative to the chroot directory.
If you wish to use \fB\-\-import\fR, you must first specify a
\fB\-\-chroot\fR directory.
If the import is unsuccessful, \fIdarkstat\fR will start with an empty
database.
.\"
.TP
.BI \-\-export " filename"
On shutdown, and upon receiving SIGUSR1, export the in-memory database
to the named file, relative to the chroot directory.
If you wish to use \fB\-\-export\fR, you must first specify a
\fB\-\-chroot\fR directory, and it must be writeable by the
\fIdarkstat\fR user.
A writeable chroot has security implications - if you are uncomfortable
with this, do not use the \fB\-\-export\fR functionality.
.\"
.TP
.BI \-\-pidfile " filename"
.RS
Creates a file containing the process ID of \fIdarkstat\fR.
This file will be unlinked upon clean shutdown.
As with all pidfiles, if \fIdarkstat\fR dies uncleanly, a stale pidfile
can be left over.

For example, start \fIdarkstat\fR with:
.IP
darkstat \-i fxp0 \-\-chroot /var/run/darkstat \-\-pidfile darkstat.pid
.PP
And stop with:
.IP
kill `cat /var/run/darkstat/darkstat.pid`
.PP
By default,
.BR kill (1)
will send SIGTERM, which will cause \fIdarkstat\fR to shut down cleanly.
.RE
.\"
.TP
.BI \-\-hosts\-max " count"
The maximum number of hosts that will be kept in the hosts table.
This is used to limit how much accounting data will be kept in memory.
The number of
.BI \-\-hosts\-max
must be greater than
.BI \-\-hosts\-keep
.\"
.TP
.BI \-\-hosts\-keep " count"
When the hosts table hits
.BI \-\-hosts\-max
and traffic is seen from a new host, we clean out the hosts table,
keeping only the top
.BI \-\-hosts\-keep
number of hosts, sorted by total traffic.
.\"
.TP
.BI \-\-ports\-max " count"
The maximum number of ports that will be tracked for each host.
This is used to limit how much accounting data will be kept in memory.
The number of
.BI \-\-ports\-max
must be greater than
.BI \-\-ports\-keep
.\"
.TP
.BI \-\-ports\-keep " count"
When a ports table fills up, this many ports are kept and the rest are
discarded.
.\"
.TP
.BI \-\-highest\-port " port"
Ports that are numerically higher than this will not appear in the
per-host ports tables, although their traffic will still be accounted
for.
This can be used to hide ephemeral ports.
By default, all ports are tracked.
.\"
.TP
.BI \-\-hexdump
Show hex dumps of received traffic.
This is only for debugging, and implies \fB\-\-verbose\fR and
\fB\-\-no\-daemon\fR.
.\"
.\" --------------------------------------------------------------------
.SH USAGE EXAMPLES
To gather statistics on the
.I fxp0
interface:
.IP
darkstat \-i fxp0
.PP
.\"
We want to account for traffic on the Internet-facing interface,
but only serve web pages to our private local network where we have the
IP address 192.168.0.1:
.IP
darkstat \-i fxp0 \-b 192.168.0.1
.PP
.\"
We want to serve web pages on the standard HTTP port:
.IP
darkstat \-i fxp0 \-p 80
.PP
.\"
We are on Optus (cable) and don't want to account for the constant ARP
traffic we are receiving:
.IP
darkstat \-i fxp0 \-f "not arp"
.PP
.\"
We only want to account for SSH traffic:
.IP
darkstat \-i fxp0 \-f "port 22"
.PP
.\"
We don't want to account for traffic between internal IPs:
.IP
darkstat \-i fxp0 \-f "not (src net 192.168.0 and dst net 192.168.0)"
.PP
.\"
(For a full reference on filter syntax, refer to the
.BR tcpdump (1)
manpage)
.PP
.\"
We have a network consisting of a gateway server (192.168.1.1) and a few
workstations (192.168.1.2, 192.168.1.3, etc.) and we want to graph all
traffic entering and leaving the local network, not just the gateway
server (which is running \fIdarkstat\fR):
.IP
darkstat \-i fxp0 \-l 192.168.1.0/255.255.255.0
.PP
.\"
On some systems, we can't capture on a "decoded" interface but
only on \fInas0\fR which returns PPPoE encapsulated packets.
Do PPPoE decoding, and override the local IP manually since it
cannot be automatically detected.
Note the /32 netmask:
.IP
darkstat \-i nas0 \-\-pppoe \-l 192.168.1.1/255.255.255.255
.\"
.SH SIGNALS
To shut
.I darkstat
down cleanly, send a SIGTERM or SIGINT signal to the
.I darkstat
parent process.
.PP
Sending the SIGUSR1 signal will cause \fIdarkstat\fR to empty out its
in-memory database.
If an \fB\-\-export\fR file was set, it will first save the database to
file.
.PP
.\"
.SH FREQUENTLY ASKED QUESTIONS
.SS How many bytes does each bar on the graph represent?
Hover your mouse cursor over a bar and you should get a tooltip
saying exactly how many bytes in and out the bar represents.
.\"
.SS Why aren't there labels / tics / a scale on the graphs?
Because implementing them is hard.
And doing so \fIcorrectly\fR, and in a way that works across all
browsers, looks pretty much impossible.

I might attempt it some day.
In the meantime, patches would be gladly accepted.
.\"
.SS Why are the graphs blank?  All the bars are zero.
The graphs only show traffic in/out of the local host, which is
determined by getting the IP address of the interface you're sniffing
on.

You can use the \fB\-l\fR argument to override the local address for
accounting purposes.
You can also use it to do accounting for a whole subnet by specifying
an appropriate netmask.
.\"
.SH SEE ALSO
.BR tcpdump (1)
.\"
.SH HISTORY
.I darkstat
was written in 2001, largely as a result of a certain Australian
cable Internet provider introducing a 3GB monthly traffic limit.
.\"
.SH AUTHORS
Emil Mikulic and others. (see the AUTHORS file)
.\"
.SH WEBSITE
http://dmr.ath.cx/net/darkstat/


NEWS
darkstat v3.0.540 and earlier defaulted to running in the foreground
and had a "-d" commandline argument to get darkstat to daemonize,
detach from the controlling terminal, and run in the background.

Since r540, the default has been inverted.  darkstat will daemonize by
default.

Since r626, this behaviour can be suppressed with the "--debug"
commandline argument, to force darkstat to stay in the foreground for
debugging purposes.

The "-d" argument is no longer recognized, and will prevent darkstat
from starting, so make sure you adjust any startup scripts you may have.

After 3.0.708, --debug was split into --verbose and --no-daemon.

Since v3.0.694, darkstat is able to save its internal database into a
file and reload it on startup.  The file format is, by design,
incompatible with the format from darkstat v2.


darkstat.c
/* darkstat 3
 * copyright (c) 2001-2010 Emil Mikulic.
 *
 * darkstat.c: signals, cmdline parsing, program body.
 *
 * You may use, modify and redistribute this file under the terms of the
 * GNU General Public License version 2. (see COPYING.GPL)
 */

#include "darkstat.h"
#include "acct.h"
#include "cap.h"
#include "conv.h"
#include "daylog.h"
#include "db.h"
#include "dns.h"
#include "http.h"
#include "hosts_db.h"
#include "localip.h"
#include "ncache.h"
#include "pidfile.h"

#include "err.h"
#include <arpa/inet.h>
#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <pcap.h>

#include "now.h"
time_t now;

#ifndef INADDR_NONE
# define INADDR_NONE (-1) /* Solaris */
#endif

/* --- Signal handling --- */
static volatile int running = 1;
static void sig_shutdown(int signum _unused_) { running = 0; }

static volatile int reset_pending = 0;
static void sig_reset(int signum _unused_) { reset_pending = 1; }

/* --- Commandline parsing --- */
static unsigned long
parsenum(const char *str, unsigned long max /* 0 for no max */)
{
   unsigned long n;
   char *end;

   errno = 0;
   n = strtoul(str, &end, 10);
   if (*end != '\0')
      errx(1, "\"%s\" is not a valid number", str);
   if (errno == ERANGE)
      errx(1, "\"%s\" is out of range", str);
   if ((max != 0) && (n > max))
      errx(1, "\"%s\" is out of range (max %lu)", str, max);
   return n;
}

const char *interface = NULL;
static void cb_interface(const char *arg) { interface = arg; }

const char *capfile = NULL;
static void cb_capfile(const char *arg) { capfile = arg; }

int want_snaplen = -1;
static void cb_snaplen(const char *arg) { want_snaplen = parsenum(arg, 0); }

int want_pppoe = 0;
static void cb_pppoe(const char *arg _unused_) { want_pppoe = 1; }

static void cb_syslog(const char *arg _unused_) { want_syslog = 1; }

static void cb_verbose(const char *arg _unused_) { want_verbose = 1; }

int want_daemonize = 1;
static void cb_no_daemon(const char *arg _unused_) { want_daemonize = 0; }

int want_promisc = 1;
static void cb_no_promisc(const char *arg _unused_) { want_promisc = 0; }

int want_dns = 1;
static void cb_no_dns(const char *arg _unused_) { want_dns = 0; }

int want_macs = 1;
static void cb_no_macs(const char *arg _unused_) { want_macs = 0; }

int want_lastseen = 1;
static void cb_no_lastseen(const char *arg _unused_) { want_lastseen = 0; }

unsigned short bindport = 667;
static void cb_port(const char *arg) { bindport = parsenum(arg, 65536); }

in_addr_t bindaddr = INADDR_ANY;
static void cb_bindaddr(const char *arg)
{
   bindaddr = inet_addr(arg);
   if (bindaddr == (in_addr_t)INADDR_NONE)
      errx(1, "malformed address \"%s\"", arg);
}

const char *filter = NULL;
static void cb_filter(const char *arg) { filter = arg; }

static void cb_local(const char *arg) { acct_init_localnet(arg); }

const char *chroot_dir = NULL;
static void cb_chroot(const char *arg) { chroot_dir = arg; }

const char *privdrop_user = NULL;
static void cb_user(const char *arg) { privdrop_user = arg; }

const char *daylog_fn = NULL;
static void cb_daylog(const char *arg)
{
   if (chroot_dir == NULL)
      errx(1, "the daylog file is relative to the chroot.\n"
      "You must specify a --chroot dir before you can use --daylog.");
   else
      daylog_fn = arg;
}

const char *import_fn = NULL;
static void cb_import(const char *arg)
{
   if (chroot_dir == NULL)
      errx(1, "the import file is relative to the chroot.\n"
      "You must specify a --chroot dir before you can use --import.");
   else
      import_fn = arg;
}

const char *export_fn = NULL;
static void cb_export(const char *arg)
{
   if ((chroot_dir == NULL) && (capfile == NULL))
      errx(1, "the export file is relative to the chroot.\n"
      "You must specify a --chroot dir before you can use --export.");
   else
      export_fn = arg;
}

static const char *pid_fn = NULL;
static void cb_pidfile(const char *arg)
{
   if (chroot_dir == NULL)
      errx(1, "the pidfile is relative to the chroot.\n"
      "You must specify a --chroot dir before you can use --pidfile.");
   else
      pid_fn = arg;
}

unsigned int hosts_max = 1000;
static void cb_hosts_max(const char *arg)
{ hosts_max = parsenum(arg, 0); }

unsigned int hosts_keep = 500;
static void cb_hosts_keep(const char *arg)
{ hosts_keep = parsenum(arg, 0); }

unsigned int ports_max = 200;
static void cb_ports_max(const char *arg)
{ ports_max = parsenum(arg, 65536); }

unsigned int ports_keep = 30;
static void cb_ports_keep(const char *arg)
{ ports_keep = parsenum(arg, 65536); }

unsigned int highest_port = 65535;
static void cb_highest_port(const char *arg)
{ highest_port = parsenum(arg, 65535); }

int wait_secs = -1;
static void cb_wait_secs(const char *arg)
{ wait_secs = parsenum(arg, 0); }

int want_hexdump = 0;
static void cb_hexdump(const char *arg _unused_) { want_hexdump = 1; }

/* --- */

struct cmdline_arg {
   const char *name, *arg_name; /* NULL arg_name means unary */
   void (*callback)(const char *arg);
   int num_seen;
};

static struct cmdline_arg cmdline_args[] = {
   {"-i",             "interface",       cb_interface,    0},
   {"-r",             "file",            cb_capfile,      0},
   {"--snaplen",      "bytes",           cb_snaplen,      0},
   {"--pppoe",        NULL,              cb_pppoe,        0},
   {"--syslog",       NULL,              cb_syslog,       0},
   {"--verbose",      NULL,              cb_verbose,      0},
   {"--no-daemon",    NULL,              cb_no_daemon,    0},
   {"--no-promisc",   NULL,              cb_no_promisc,   0},
   {"--no-dns",       NULL,              cb_no_dns,       0},
   {"--no-macs",      NULL,              cb_no_macs,      0},
   {"--no-lastseen",  NULL,              cb_no_lastseen,  0},
   {"-p",             "port",            cb_port,         0},
   {"-b",             "bindaddr",        cb_bindaddr,     0},
   {"-f",             "filter",          cb_filter,       0},
   {"-l",             "network/netmask", cb_local,        0},
   {"--chroot",       "dir",             cb_chroot,       0},
   {"--user",         "username",        cb_user,         0},
   {"--daylog",       "filename",        cb_daylog,       0},
   {"--import",       "filename",        cb_import,       0},
   {"--export",       "filename",        cb_export,       0},
   {"--pidfile",      "filename",        cb_pidfile,      0},
   {"--hosts-max",    "count",           cb_hosts_max,    0},
   {"--hosts-keep",   "count",           cb_hosts_keep,   0},
   {"--ports-max",    "count",           cb_ports_max,    0},
   {"--ports-keep",   "count",           cb_ports_keep,   0},
   {"--highest-port", "port",            cb_highest_port, 0},
   {"--wait",         "secs",            cb_wait_secs,    0},
   {"--hexdump",      NULL,              cb_hexdump,      0},
   {NULL,             NULL,              NULL,            0}
};

static void
pad(const int width)
{
   int i;
   for (i=0; i<width; i++) printf(" ");
}

/*
 * We autogenerate the usage statement from the cmdline_args data structure.
 */
static void
usage(void)
{
   int width, first;
   struct cmdline_arg *arg;

   printf(PACKAGE_STRING " (built with libpcap %d.%d)\n\n",
      PCAP_VERSION_MAJOR, PCAP_VERSION_MINOR);

   width = printf("usage: darkstat ");
   first = 1;

   for (arg = cmdline_args; arg->name != NULL; arg++) {
      if (first) first = 0; else pad(width);
      printf("[ %s", arg->name);
      if (arg->arg_name != NULL) printf(" %s", arg->arg_name);
      printf(" ]\n");
   }
   printf("\n"
"Please refer to the darkstat(8) manual page for further\n"
"documentation and usage examples.\n");
}

static void
parse_sub_cmdline(const int argc, char * const *argv)
{
   struct cmdline_arg *arg;

   if (argc == 0) return;
   for (arg = cmdline_args; arg->name != NULL; arg++)
      if (strcmp(argv[0], arg->name) == 0) {
         if ((arg->arg_name != NULL) && (argc == 1)) {
            fprintf(stderr,
               "error: argument \"%s\" requires parameter \"%s\"\n",
               arg->name, arg->arg_name);
            usage();
            exit(EXIT_FAILURE);
         }
         if (arg->num_seen > 0) {
            fprintf(stderr,
               "error: already specified argument \"%s\"\n",
               arg->name);
            usage();
            exit(EXIT_FAILURE);
         }

         arg->num_seen++;
         if (arg->arg_name == NULL) {
            arg->callback(NULL);
            parse_sub_cmdline(argc-1, argv+1);
         } else {
            arg->callback(argv[1]);
            parse_sub_cmdline(argc-2, argv+2);
         }
         return;
      }

   fprintf(stderr, "error: illegal argument: \"%s\"\n", argv[0]);
   usage();
   exit(EXIT_FAILURE);
}

static void
parse_cmdline(const int argc, char * const *argv)
{
   if (argc < 1) {
      /* Not enough args. */
      usage();
      exit(EXIT_FAILURE);
   }

   parse_sub_cmdline(argc, argv);

   /* start syslogging as early as possible */
   if (want_syslog) openlog("darkstat", LOG_NDELAY | LOG_PID, LOG_DAEMON);

   /* some default values */
   if (chroot_dir == NULL) chroot_dir = CHROOT_DIR;
   if (privdrop_user == NULL) privdrop_user = PRIVDROP_USER;

   /* sanity check args */
   if ((interface == NULL) && (capfile == NULL))
      errx(1, "must specify either interface (-i) or capture file (-r)");

   if ((interface != NULL) && (capfile != NULL))
      errx(1, "can't specify both interface (-i) and capture file (-r)");

   if ((hosts_max != 0) && (hosts_keep >= hosts_max)) {
      hosts_keep = hosts_max / 2;
      warnx("reducing --hosts-keep to %u, to be under --hosts-max (%u)",
         hosts_keep, hosts_max);
   }
   verbosef("max %u hosts, cutting down to %u when exceeded",
      hosts_max, hosts_keep);

   if ((ports_max != 0) && (ports_keep >= ports_max)) {
      ports_keep = ports_max / 2;
      warnx("reducing --ports-keep to %u, to be under --ports-max (%u)",
         ports_keep, ports_max);
   }
   verbosef("max %u ports per host, cutting down to %u when exceeded",
      ports_max, ports_keep);

   if (want_hexdump && !want_verbose) {
      want_verbose = 1;
      verbosef("--hexdump implies --verbose");
   }

   if (want_hexdump && want_daemonize) {
      want_daemonize = 0;
      verbosef("--hexdump implies --no-daemon");
   }
}

static void
run_from_capfile(void)
{
   graph_init();
   hosts_db_init();
   cap_from_file(capfile, filter);
   cap_stop();
   if (export_fn != NULL) db_export(export_fn);
   hosts_db_free();
   graph_free();
   verbosef("Total packets: %qu, bytes: %qu", total_packets, total_bytes);
}

/* --- Program body --- */
int
main(int argc, char **argv)
{
   test_64order();
   parse_cmdline(argc-1, argv+1);

   if (capfile) {
      /*
       * This is very different from a regular run against a network
       * interface.
       */
      run_from_capfile();
      return 0;
   }

   /* must verbosef() before first fork to init lock */
   verbosef("starting up");
   if (pid_fn) pidfile_create(chroot_dir, pid_fn, privdrop_user);

   if (want_daemonize) {
      verbosef("daemonizing to run in the background!");
      daemonize_start();
      verbosef("I am the main process");
   }
   if (pid_fn) pidfile_write_close();

   /* do this first as it forks - minimize memory use */
   if (want_dns) dns_init(privdrop_user);
   cap_init(interface, filter, want_promisc); /* needs root */
   http_init(bindaddr, bindport, /*maxconn=*/ -1); /* low ports need root */
   ncache_init(); /* must do before chroot() */

   privdrop(chroot_dir, privdrop_user);

   /* Don't need root privs for these: */
   now = time(NULL);
   if (daylog_fn != NULL) daylog_init(daylog_fn);
   graph_init();
   hosts_db_init();
   if (import_fn != NULL) db_import(import_fn);
   localip_init(interface);

   if (signal(SIGTERM, sig_shutdown) == SIG_ERR)
      errx(1, "signal(SIGTERM) failed");
   if (signal(SIGINT, sig_shutdown) == SIG_ERR)
      errx(1, "signal(SIGINT) failed");
   if (signal(SIGUSR1, sig_reset) == SIG_ERR)
      errx(1, "signal(SIGUSR1) failed");

   verbosef("entering main loop");
   daemonize_finish();

   while (running) {
      int select_ret, max_fd = -1, use_timeout = 0;
      struct timeval timeout;
      fd_set rs, ws;

      now = time(NULL);

      if (reset_pending) {
         if (export_fn != NULL) db_export(export_fn); /* FIXME: USR2? */
         hosts_db_reset();
         graph_reset();
         reset_pending = 0;
      }

      FD_ZERO(&rs);
      FD_ZERO(&ws);

      cap_fd_set(&rs, &max_fd, &timeout, &use_timeout);
      http_fd_set(&rs, &ws, &max_fd, &timeout, &use_timeout);

      select_ret = select(max_fd+1, &rs, &ws, NULL,
         (use_timeout) ? &timeout : NULL);

      if ((select_ret == 0) && (!use_timeout))
            errx(1, "select() erroneously timed out");

      if (select_ret == -1) {
         if (errno == EINTR)
            continue;
         else
            err(1, "select()");
      }
      else {
         graph_rotate();
         cap_poll(&rs);
         dns_poll();
         http_poll(&rs, &ws);
      }
   }

   verbosef("shutting down");
   verbosef("pcap stats: %u packets received, %u packets dropped",
      pkts_recv, pkts_drop);
   cap_stop();
   dns_stop();
   if (export_fn != NULL) db_export(export_fn);
   hosts_db_free();
   graph_free();
   if (daylog_fn != NULL) daylog_free();
   ncache_free();
   if (pid_fn) pidfile_unlink();
   verbosef("shut down");
   return (EXIT_SUCCESS);
}

/* vim:set ts=3 sw=3 tw=78 expandtab: */
